### PR TITLE
gravel: add network controller and API endpoints

### PR DESCRIPTION
**Pull request by tserong**
This adds the ability for gravel to read and write network interface configuration, nameservers and routes.  We've added an /api/local/network/config endpoint.  GET this, and you'll retrieve a dict of interfaces, a list of nameservers and a list of static routes.  In the simple case for a single NIC that's configured for DHCP, this'll give you:

```json
{
  "interfaces": {
     "eth0": {
      "name": "eth0",
      "config": {
        "bootproto": "dhcp",
        "ipaddr": "",
        "bonding_slaves": []
      }
    }
  },
  "nameservers": [],
  "routes": []
}
```

POST back to that same URL to apply a different config.  The config exposed is deliberately a subset of what's possible to specify in /etc/sysconfig/network/{config,routes,ifcfg-*,ifroute-*} files.  For 'interfaces', bootproto can be either "dhcp" or "static".  If it's "static", you also need to specify "ipaddr", and that needs to include the netmask or equivalent on the end (e.g.: "192.168.121.1/24"). If you want to create a bond, add a new interface to that dict and set bonding_slaves to a list of the physical interfaces to bond.  If the list is empty, it's just a normal interface.

'nameservers' is just a list of IP addresses, e.g.: ["8.8.8.8", ...]

'routes' is a list of dicts, each of which needs to include 'destination', 'gateway' and 'interface' as described in `man ifroute`.

When you apply the config, all network services are restarted.  Be particularly careful when applying a new config to the interface via which you're accessing the API.  Switch from DHCP to static IP and forget to configure DNS and/or a default route, and you can get yourself in trouble.  Likewise if you change the IP address to be on some other network.

Note that for NICs we've got an InterfaceModel which includes an InterfaceConfigModel.  This was based on Joao's work in https://github.com/aquarist-labs/aquarium/pull/603, the idea being that we may also want to add stats and other data to InterfaceModel. OTOH, that may duplicate information we already get from cephadm, so we might want to consider collapsing this instead.

Signed-off-by: Tim Serong <tserong@suse.com>

## Checklist
- [ ] References issues, create if required
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins run tumbleweed`
- `jenkins run leap`
- `jenkins run ubuntu`

</details>

### CODE REVIEW

**Comment by tserong** [comment]
I've now also added the ability to pass network config in to the /nodes/deployment/start and /nodes/join endpoints. I've tested this by manually hitting those endpoints with an appropriate request, e.g. for starting the deployment:

```
{
  "hostname": "node1",
  "network": {
    "interfaces": {
      "eth0": { 
        "name": "eth0",
        "config": {
          "bootproto": "static",
          "ipaddr": "192.168.121.89/24",
          "bonding_slaves": []
        }
      }
    },
    "nameservers": [
      "192.168.121.1"
    ],
    "routes": [
      {
        "destination": "default",
        "gateway": "192.168.121.1",
        "interface": "-"
      }
    ]
  },
  "ntpaddr": "pool.ntp.org",
  "registry": { 
    "registry": "registry.opensuse.org",
    "secure": true,
    "image": "filesystems/ceph/master/upstream/images/opensuse/ceph/ceph:latest"
  }
}
```

In the above case I'm applying a static IP address, but was careful to assign the same address as was already assigned by DHCP. If I'd chosen a different IP address, the browser would have lost its connection to the UI. This suggests that once network config is added to the setup wizard, we need some means of issuing a redirect if the IP address you're accessing aquarium on changes.

Anyway: the backend bits basically work, and there should, I hope, be enough in place, that the frontend bits can be done next if this PR passes review.

**Comment by tserong** [comment]
(Happy to squash these commits, BTW)

File: src/gravel/controllers/resources/network.py
Comment (jhesketh) [inline src/gravel/controllers/resources/network.py:104]: Is it correct to assume the config file in that position will be up to date? Would we be better off using `ip addr` and associated utils to get the information? Using the latter would also likely make it easier to move between wicked and networkmanager.

File: src/gravel/controllers/resources/network.py
Comment (jhesketh) [inline src/gravel/controllers/resources/network.py:158]: same question vs using other utilities (or /etc/resolv.conf etc).

File: src/gravel/controllers/resources/network.py
Comment (jhesketh) [inline src/gravel/controllers/resources/network.py:171]: also from `ip route`?

File: src/gravel/controllers/resources/network.py
Comment (jhesketh) [inline src/gravel/controllers/resources/network.py:242]: The `ip` utility can make these changes (afaik).

**Comment by jhesketh** [comment]
Generally looks good to me, but wondering about the best way to read and apply the config (see inline). Sorry if it's already been discussed etc.

Your PR description would also make some helpful docs if you wanted to chuck that into git somewhere until we start writing real documentation.

**Review by jhesketh — DISMISSED** [review]

**Comment by tserong** [comment]
> Generally looks good to me, but wondering about the best way to read and apply the config (see inline). Sorry if it's already been discussed etc.

I forget whether or not this got discussed in detail anywhere, sorry :-) but the assumptions I was working from were:

- We'll always by using wicked, not networkmanager (because the former is the default on SUSE distros for servers, the latter for laptops/desktops)
- We only want to expose interfaces that physical exist (/sys/class/net/*/device is present) or are persistently configured (/etc/sysconfig/network/ifcfg-*)
- Nobody should be making changes to the network manually, outside of aquarium (nothing ephemeral that we'd need to pick up from `ip` invocations).

File: src/gravel/controllers/resources/network.py
Comment (tserong) [reply src/gravel/controllers/resources/network.py:158]: With the default setup, /etc/resolv.conf is dynamically generated by netconfig based on variables in /etc/sysconfig/network/config, so it should be OK like this, provided my various assumptions are correct - see my other comment on that.

**Comment by tserong** [comment]
Yikes ... resolving conflicts via the github web UI merges main into this branch. I don't like it - I'm gonna go rebase instead like I usually do.

**Review by jecluis — DISMISSED** [review]
LGTM but we might want to update the config with new keys from the dict in `_update_config()`. I'll leave that to @tserong 's consideration. :)

**Comment by tserong** [comment]
@jecluis I've got it adding new keys now in `_update_config()`. We may not strictly have needed that with the default setup, but it was an oversight on my part (thank you!), and realising that I figured it'd find a way to bite us later somehow if I didn't fix it ;-)

Mind a quick re-review? Also: want me to squash all these commits, or just merge as-is once approved?

**Review by jecluis — APPROVED** [review]

**Comment by jecluis** [comment]
@tserong good to go if you are happy with it.

**Pull request merged**
